src/components/DocumentRenderer/QrCode/QrCodeContainer.tsx

```diff
@@ -0,0 +1,163 @@
+  useCallback
+} from "react";
+import { StyleSheet, View, Text } from "react-native";
+import {
+  color,
+  borderRadius,
+  shadow,
+  size,
+  fontSize
+} from "../../../common/styles";
+import { QrCode } from "./QrCode";
+import { DocumentProperties, DocumentObject } from "../../../types";
+import { useQrGenerator } from "../../../common/hooks/useQrGenerator";
+import { useNetworkContext } from "../../../context/network";
+import HumanizeDuration from "humanize-duration";
+import { useCountdown } from "../../../common/hooks/useCountdown";
+
+const styles = StyleSheet.create({
+  wrapper: {
+    backgroundColor: color("grey", 0),
+    borderRadius: borderRadius(2),
+    ...shadow(2)
+  },
+  qrCodeWrapper: {
+    aspectRatio: 1
+  },
+  feedback: {
+    marginTop: -size(2)
+  },
+  expiryText: {
+    fontSize: fontSize(-3),
+    marginHorizontal: size(3),
+    marginBottom: size(2),
+    textAlign: "center"
+  },
+  offlineBanner: {
+    backgroundColor: color("red", 10),
+    paddingVertical: size(2),
+    paddingHorizontal: size(3),
+    borderRadius: borderRadius(2),
+    borderTopLeftRadius: 0,
+    borderTopRightRadius: 0
+  },
+  offlineBannerHeader: {
+    fontSize: fontSize(-2),
+    color: color("red", 30),
+    fontWeight: "bold",
+    marginBottom: size(0.5)
+  },
+  offlineBannerText: {
+    fontSize: fontSize(-3),
+    color: color("red", 30)
+  }
+});
+
+const updateDocumentQr = async (
+  qrCode: { url: string; expiry?: number },
+  document: { atomicUpdate: DocumentObject["atomicUpdate"] }
+): Promise<void> => {
+  const updateFunction = (oldData: DocumentProperties): DocumentProperties => {
+    oldData.qrCode = { ...qrCode };
+    return oldData;
+  };
+  await document?.atomicUpdate(updateFunction);
+};
+
+export interface QrCodeContainerRef {
+  triggerGenerateQr: () => void;
+}
+
+interface QrCodeContainer {
+  document: DocumentProperties & Pick<DocumentObject, "atomicUpdate">;
+  refreshPaused: boolean;
+}
+
+// eslint-disable-next-line react/display-name
+export const QrCodeContainer = forwardRef<QrCodeContainerRef, QrCodeContainer>(
+  ({ document, refreshPaused }, ref) => {
+    const { isConnected } = useNetworkContext();
+
+    const { qrCode, qrCodeLoading, generateQr } = useQrGenerator(
+      document.qrCode
+    );
+
+    // Updates the stored QR code in the DB
+    useEffect(() => {
+      updateDocumentQr(qrCode, document);
+    }, [qrCode, document]);
+
+    const [expiry, setExpiry] = useState<number>();
+    const { secondsLeft, startCountdown } = useCountdown();
+
+    // Starts the countdown based on the expiry of the QR code
+    useEffect(() => {
+      if (qrCode.expiry !== undefined) {
+        setExpiry(qrCode.expiry);
+        startCountdown(Math.round((qrCode.expiry - Date.now()) / 1000));
+      }
+    }, [qrCode.expiry, startCountdown]);
+
+    // Resets the QR code URL when it expires
+    useEffect(() => {
+      if (secondsLeft === 0) {
+        updateDocumentQr({ url: "" }, document);
```

Comment:
Why do we need to invalidate this when it expires? Since expiry is already there wouldn’t it make sense that we ignore it and then the next time we fetch it just doesn’t show those that has expired?

Reply:
As discussed, we won't reset the url when it expires, will let the consumer handle the visual logic